internal/security/proxy/consumer.go

```diff
@@ -0,0 +1,237 @@
+	resp, err := c.Connect.GetHTTPClient().Do(req)
+	if err != nil {
+		e := fmt.Sprintf("failed to associate consumer %s for with group %s with error %s", c.Name, g, err.Error())
+		lc.Info(e)
+		return errors.New(e)
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusOK, http.StatusCreated, http.StatusConflict:
+		lc.Info(fmt.Sprintf("successful to associate consumer %s with group %s", c.Name, g))
+		break
+	default:
+		b, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return err
+		}
+		e := fmt.Sprintf("failed to associate consumer %s with group %s with error %s,%s", c.Name, g, resp.Status, string(b))
+		lc.Error(e)
+		return errors.New(e)
+	}
+	return nil
+}
+
+func (c *Consumer) CreateToken() (string, error) {
+	switch c.Cfg.GetProxyAuthMethod() {
+	case "jwt":
+		lc.Info("autheticate the user with jwt authentication.")
+		return c.createJWTToken()
+	case "oauth2":
+		lc.Info("authenticate the user with oauth2 authentication.")
+		return c.createOAuth2Token()
+	default:
+		e := fmt.Sprintf("unknown authentication method provided: %s", c.Cfg.GetProxyAuthMethod())
+		lc.Info(e)
+		return "", errors.New(e)
+	}
+}
+
+func (c *Consumer) createJWTToken() (string, error) {
+	jwtCred := JWTCred{}
+	s := sling.New().Set("Content-Type", "application/x-www-form-urlencoded")
+	req, err := s.New().Get(c.Connect.GetProxyBaseURL()).Post(fmt.Sprintf("consumers/%s/jwt", c.Name)).Request()
+	resp, err := c.Connect.GetHTTPClient().Do(req)
+	if err != nil {
+		e := fmt.Sprintf("failed to create jwt token for consumer %s with error %s", c.Name, err.Error())
+		return "", errors.New(e)
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusOK, http.StatusCreated, http.StatusConflict:
+		if err = json.NewDecoder(resp.Body).Decode(&jwtCred); err != nil {
+			return "", err
+		}
+		lc.Info(fmt.Sprintf("successful on retrieving JWT credential for consumer %s", c.Name))
+
+		// Create the Claims
+		claims := KongJWTClaims{
+			jwtCred.Key,
+			c.Name,
+			jwt.StandardClaims{
+				Issuer: EdgeXService,
+			},
+		}
+
+		token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
+		return token.SignedString([]byte(jwtCred.Secret))
+	default:
+		e := fmt.Sprintf("failed to create JWT for consumer %s with errorCode %d", c.Name, resp.StatusCode)
+		lc.Info(e)
+		return "", errors.New(e)
+	}
+}
+
+//curl -X POST "http://localhost:8001/consumers/user123/oauth2" -d "name=www.edgexfoundry.org" --data "client_id=user123" -d "client_secret=user123"  -d "redirect_uri=http://www.www.edgexfoundry.org/"
+//curl -k -v https://localhost:8443/{service}/oauth2/token -d "client_id=user123" -d "grant_type=client_credentials" -d "client_secret=user123" -d "scope=email"
+func (c *Consumer) createOAuth2Token() (string, error) {
+
+	url := fmt.Sprintf("http://%s:%s/", c.Cfg.GetProxyServerName(), c.Cfg.GetProxyServerPort())
+
+	tr := &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
```

Comment:
the 2nd request for OAuth token requires that it has to be requested on https but we are using self-signed cert so the HTTP request will be rejected without setting inscureskipverify as true.

Reply:
but kong could be configured to not use a self-signed certificate, so we shouldn't be hard-coding InsecureSkipVerify here

Reply:
Updated the code base so that InsecureSkipVerify will be configurable based on the command line parameters, same as other places where we pass this parameter.